docscan/client.go

```diff
@@ -0,0 +1,94 @@
+package docscan
+
+import (
+	"crypto/rsa"
+	"net/http"
+	"os"
+
+	"github.com/getyoti/yoti-go-sdk/v3/cryptoutil"
+	"github.com/getyoti/yoti-go-sdk/v3/docscan/service"
+	"github.com/getyoti/yoti-go-sdk/v3/docscan/session/create"
+	"github.com/getyoti/yoti-go-sdk/v3/docscan/session/retrieve"
+	"github.com/getyoti/yoti-go-sdk/v3/docscan/supported"
+	"github.com/getyoti/yoti-go-sdk/v3/media"
+	"github.com/getyoti/yoti-go-sdk/v3/requests"
+)
+
+// Client is responsible for setting up test data in the sandbox instance.
+type Client struct {
+	// SDK ID. This can be found in the Yoti Hub after you have created and activated an application.
+	SdkID string
+	// Private Key associated for your application, can be downloaded from the Yoti Hub.
+	Key *rsa.PrivateKey
+	// Mockable HTTP Client Interface
+	HTTPClient requests.HttpClient
+	// API URL to use. This is not required, and a default will be set if not provided.
+	apiURL string
+}
+
+// NewClient constructs a Client object
+func NewClient(sdkID string, key []byte) (*Client, error) {
+	decodedKey, err := cryptoutil.ParseRSAKey(key)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		SdkID: sdkID,
+		Key:   decodedKey,
+	}, err
+}
+
+// OverrideAPIURL overrides the default API URL for this Yoti Client
+func (client *Client) OverrideAPIURL(apiURL string) {
+	client.apiURL = apiURL
+}
+
+func (client *Client) getAPIURL() string {
+	if client.apiURL == "" {
+		if value, exists := os.LookupEnv("YOTI_DOC_SCAN_API_URL"); exists && value != "" {
+			client.apiURL = value
+		} else {
+			client.apiURL = "https://api.yoti.com/sandbox/idverify/v1"
+		}
+	}
+	return client.apiURL
+}
+
+func (client *Client) getHTTPClient() requests.HttpClient {
+	if client.HTTPClient != nil {
+		return client.HTTPClient
+	}
+	return http.DefaultClient
+}
+
+// CreateSession creates a Doc Scan session using the supplied session specification
+func (client *Client) CreateSession(sessionSpec *create.SessionSpecification) (*create.SessionResult, error) {
+	return service.CreateSession(client.getHTTPClient(), client.SdkID, client.Key, client.getAPIURL(), sessionSpec, nil)
```

Comment:
It might be neater to inject `client.getHTTPClient()`, `client.SdkID`, `client.Key` and `client.getAPIURL()` into the service, then call with the parameters related to this method: `service.CreateSession(sessionSpec)`

This might not be easy to do neatly this as we have to consider `OverrideAPIURL` and ensure this is propagated to the service.

I think it might make sense to drop `docscan/service/session.go` and move the request builder code into the client

Reply:
Moved into the client in 4e7b2cc. Partly to match Doc Scan Sandbox, and also to mitigate the chance of breaking things if we change something which would now be internal do docscan.client